### PR TITLE
AC-451 :Fixed app crash on opening Form entry

### PR DESCRIPTION
**Pull request by ivary43**
Issue Link : https://issues.openmrs.org/browse/AC-451 .

### CODE REVIEW

**Comment by ivary43** [comment]
@dkayiwa , @Sub6Resources please review.

File: openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java [hunk 1]
```diff
@@ -218,4 +222,25 @@ public static boolean validateDate(String dateString, DateTime minDate, DateTime
...
+    /**
+     * To validate the Date format
+     * Date format used is dd/mm/YYYY
+     *
+     * @param format       to check the format of the string
+     * @param dateAsString the value of raw date
+     * @return true if the date format matches the string value
+     */
+    public static boolean isValidFormat(String format, String dateAsString) {
+        Date date = null;
```
Comment (Sub6Resources) [inline openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: You should only need to use `Date date;` as the variable will by default be null.

File: openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java [hunk 2]
```diff
@@ -218,4 +222,25 @@ public static boolean validateDate(String dateString, DateTime minDate, DateTime
...
+        try {
+            SimpleDateFormat simpleDateFormat = new SimpleDateFormat(format);
+            date = simpleDateFormat.parse(dateAsString);
+            if (!dateAsString.equals(simpleDateFormat.format(date))) {
+                date = null;
+            }
+        } catch (ParseException exception) {
+            OpenMRS.getInstance().getOpenMRSLogger().w("Failed to validate date format :" + dateAsString + " caused by " + exception.toString());
+        }
+        return date != null;
```
Comment (ivary43) [inline openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: @Sub6Resources  i will be getting an error(Variable uninitialised) here if I remove that line.

Comment (Sub6Resources) [reply openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: Okay, don't worry about that then.

File: openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java (hunk 1 above)
Comment (dkayiwa) [reply openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: @ivary43 did you see the above comment?

Comment (ivary43) [reply openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: @dkayiwa yes I did.I will get an error if I don't initialise that.

Comment (dkayiwa) [reply openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: which error?

File: openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java (hunk 2 above)
Comment (ivary43) [reply openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: @dkayiwa  i will get an error here in
 ```
return data!=null;
```

Comment (dkayiwa) [reply openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: with what error message?

Comment (ivary43) [reply openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]:
```
data is not initialised.
```

Comment (dkayiwa) [reply openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: Can you claim the ticket and takes it through the process mentioned at? https://wiki.openmrs.org/display/docs/Pull+Request+Tips

Comment (ivary43) [reply openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: Did that 😄 .

Comment (dkayiwa) [reply openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: Can you take a look at this? https://wiki.openmrs.org/display/docs/Pull+Request+Tips

Comment (ivary43) [reply openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: I went through it. Is there something is missed I have done all of it requested for review, claimed it and referenced it
in the ticket.

File: openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java [hunk 3]
```diff
@@ -218,4 +222,25 @@ public static boolean validateDate(String dateString, DateTime minDate, DateTime
         }
     }
 
+    /**
+     * To validate the Date format
+     * Date format used is dd/mm/YYYY
+     *
+     * @param format       to check the format of the string
```
Comment (dkayiwa) [inline openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: to check the format of the string or the format of the string? which one is which?

Comment (ivary43) [reply openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: Sorry for the grammatical error I will fix that.Is that the only problem??

**Comment by ivary43** [comment]
@dkayiwa  please check I have made the changes.

**Comment by dkayiwa** [comment]
@ivary43 also the ticket link does not look correct.

**Comment by ivary43** [comment]
@dkayiwa I corrected that.

**Comment by dkayiwa** [comment]
Did you also see my comment on the ticket?

File: openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java [hunk 4]
```diff
@@ -218,4 +222,25 @@ public static boolean validateDate(String dateString, DateTime minDate, DateTime
         }
     }
 
+    /**
+     * To validate the Date format
```
Comment (dkayiwa) [inline openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: What does this method do?

Comment (ivary43) [reply openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: It validates the format of the date,  is it in ```DEFAULT_DATE_FORMAT``` or not .

File: openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java [hunk 5]
```diff
@@ -218,4 +222,25 @@ public static boolean validateDate(String dateString, DateTime minDate, DateTime
...
+     * Date format used is dd/mm/YYYY
+     *
+     * @param format the format to check against
+     * @param dateAsString the value of raw date
+     * @return true if the date format matches the string value
```
Comment (dkayiwa) [inline openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: return true if the given date matches the given format

File: openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java (hunk 4 above)
Comment (dkayiwa) [inline openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: Validates a date against a given format

**Comment by ivary43** [comment]
@dkayiwa  did that.

File: openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java [hunk 6]
```diff
@@ -218,4 +222,25 @@ public static boolean validateDate(String dateString, DateTime minDate, DateTime
         }
     }
 
+    /**
+     * Validates a date against a given format
+     * Date format used is dd/mm/YYYY
```
Comment (dkayiwa) [inline openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: How do you tell that this is the date format used?

Comment (ivary43) [reply openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: From logcat...i reported in the ticket too.

Comment (dkayiwa) [reply openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: Isn't it dynamically read from the user's settings?

Comment (ivary43) [reply openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: No, it's basically a server-side response which contains both types of date format ```DEFAULT_DATE_FORMAT``` and ```OPEN_MRS_RESPONSE_FORMAT``` both  only in string format so earlier it was passed to ```  public static Long convertTime(String dateAsString) ``` to this constructor which was converting the datestring by assuming it's of this format ```OPEN_MRS_RESPONSE_FORMAT``` but string type contained both type . so what i did was put a check for type of format then converted it corresponding to that format.

Comment (dkayiwa) [reply openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: This is public function which takes a format as one of the parameters. The mere fact that the format is passed in as a parameter means it is not hard coded as per your javadoc description above.

Comment (ivary43) [reply openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: @dkayiwa i have made the changes .

Comment (dkayiwa) [reply openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: This has totally changed the method. Originally, it gave one the freedom of specifying a date format. Now you have hard coded it hence making it less flexible.

Comment (ivary43) [reply openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: @dkayiwa  what are the changes you want?

Comment (dkayiwa) [reply openmrs-client/src/main/java/org/openmrs/mobile/utilities/DateUtils.java]: To remove the hard coding of the format from the javadoc

**Comment by ivary43** [comment]
@dkayiwa  i removed that.

**Comment by dkayiwa** [comment]
The ticket status shows that you are still doing more work on this.

**Comment by ivary43** [comment]
@dkayiwa please check.

**Pull request merged**